lib/web/app/transport.go

```diff
@@ -139,9 +145,58 @@ func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
 		return nil, trace.Wrap(err)
 	}
 
+	// When proxying app in leaf cluster, the app will have PublicAddr
+	// possibly unreachable to the clients connecting to the root cluster.
+	// We want to rewrite any redirects to that PublicAddr to equivalent redirect for root cluster.
+	if err = t.rewriteRedirect(resp); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
 	return resp, nil
 }
 
+// rewriteRedirect rewrites redirect to a public address of downstream app.
+func (t *transport) rewriteRedirect(resp *http.Response) error {
+	if !t.c.isRemoteApp() {
+		// Not a downstream app, short circuit.
+		return nil
+	}
+
+	if !utils.IsRedirect(resp.StatusCode) {
+		return nil
+	}
+
+	location := resp.Header.Get("Location")
+	// nothing to do without Location.
+	if location == "" {
+		return nil
+	}
+
+	// Parse the "Location" header.
+	u, err := url.Parse(location)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	host, _, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		host = u.Host
+	}
+
+	// If this is a redirect to a public addr of a leaf cluster, rewrite it.
+	// We want the rewrite to happen using our own public address.
+	if host == t.c.identity.RouteToApp.PublicAddr {
```

Comment:
Is this check really needed? Wouldn't we want to rewrite redirect location to root's address in any case?

Reply:
Redirects to other addresses - like `https://goteleport.com/` - should stay as they are.